src/bidiMapper/domains/context/scriptEvaluator.ts

```diff
@@ -0,0 +1,394 @@
+        this.#callFunctionStacktraceLineOffset
+      );
+    }
+
+    return { result: ScriptEvaluator.#cdpToBidiValue(cdpCallFunctionResult) };
+  }
+
+  async #serializeCdpExceptionDetails(
+    cdpExceptionDetails: Protocol.Runtime.ExceptionDetails,
+    lineOffset: number
+  ): Promise<Message.ExceptionResult> {
+    const callFrames = cdpExceptionDetails.stackTrace?.callFrames.map(
+      (frame) => ({
+        url: frame.url,
+        functionName: frame.functionName,
+        // As `script.evaluate` wraps call into serialization script, so
+        // `lineNumber` should be adjusted.
+        lineNumber: frame.lineNumber - lineOffset,
+        columnNumber: frame.columnNumber,
+      })
+    );
+
+    const exception = await this.serializeCdpObject(
+      // Exception should always be there.
+      cdpExceptionDetails.exception!
+    );
+
+    const text = await this.stringifyObject(cdpExceptionDetails.exception!);
+
+    return {
+      exceptionDetails: {
+        exception,
+        columnNumber: cdpExceptionDetails.columnNumber,
+        // As `script.evaluate` wraps call into serialization script, so
+        // `lineNumber` should be adjusted.
+        lineNumber: cdpExceptionDetails.lineNumber - lineOffset,
+        stackTrace: {
+          callFrames: callFrames || [],
+        },
+        text: text || cdpExceptionDetails.text,
+      },
+    };
+  }
+
+  static #cdpToBidiValue(
+    cdpValue:
+      | Protocol.Runtime.CallFunctionOnResponse
+      | Protocol.Runtime.EvaluateResponse
+  ): CommonDataTypes.RemoteValue {
+    const bidiValue = cdpValue.result.webDriverValue!;
+    if (cdpValue.result.objectId) {
+      bidiValue.objectId = cdpValue.result.objectId;
+    }
+    // This relies on the CDP to implement proper BiDi serialization, except
+    // objectIds.
+    return bidiValue as CommonDataTypes.RemoteValue;
+  }
+
+  public async scriptEvaluate(
+    expression: string,
+    awaitPromise: boolean
+  ): Promise<Script.EvaluateResult> {
+    let cdpEvaluateResult = await this.#cdpClient.Runtime.evaluate({
+      expression,
+      awaitPromise,
+      generateWebDriverValue: true,
+    });
+
+    if (cdpEvaluateResult.exceptionDetails) {
+      // Serialize exception details.
+      return await this.#serializeCdpExceptionDetails(
+        cdpEvaluateResult.exceptionDetails,
+        this.#evaluateStacktraceLineOffset
+      );
+    }
+
+    return {
+      result: ScriptEvaluator.#cdpToBidiValue(cdpEvaluateResult),
+    };
+  }
+
+  async #deserializeToCdpArg(
+    serializedValue: CommonDataTypes.LocalValue
+  ): Promise<Protocol.Runtime.CallArgument> {
+    if (`objectId` in serializedValue) {
+      return { objectId: serializedValue.objectId };
+    }
+    switch (serializedValue.type) {
+      // Primitive Protocol Value
+      // https://w3c.github.io/webdriver-bidi/#data-types-protocolValue-primitiveProtocolValue
+      case 'undefined': {
+        return { unserializableValue: 'undefined' };
```

Comment:
what does unserializableValue means? a value that cannot be deserialized?

Reply:
[`Runtime.CallArgument. unserializableValue`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#type-CallArgument) is basically a string that is evaluated instead of being JSON-deserialized.